src/transformers/trainer.py

```diff
@@ -233,25 +250,33 @@ def __init__(
         self.args = args
         # Seed must be set before instantiating the model when using model
         set_seed(self.args.seed)
-        assert (
-            model is not None or model_init is not None
-        ), "You must provide a model to use `Trainer`, either by using the `model` argument or the `model_init` argument."
-        self.model_init = model_init
         self.hp_name = None
-        if model is None and model_init is not None:
-            model = self.call_model_init()
 
-        # Model parallel
-        if model is not None and not self.args.model_parallel:
-            model = model.to(args.device)
+        if model is None:
+            if model_init is not None:
+                self.model_init = model_init
+                model = self.call_model_init()
+            else:
+                raise RuntimeError("`Trainer` requires either a `model` or `model_init` argument")
+        else:
+            if model_init is not None:
+                raise RuntimeError("`Trainer` requires either a `model` or `model_init` argument, but not both")
+            self.model_init = None
```

Comment:
This is new and should just warrant a warning (something like: "You provided a model_init and a model, your model will be replaced at the next call to `Trainer.train`") in my opinion, not an error. All examples using hyperparameter search might break for instance as they could be using both model and model_init.

Reply:
Let's please discuss the logic here, decide on whether the above logic is correct and then choose the assert or warning accordingly.

Pre-this PR logic silently ignored `model_init`.

Why would a caller provide both? If the code logic forcefully ignores `model_init` should `model` be provided then the user is being mislead into thinking that perhaps `model_init` is executed? If it's erroneous to provide both then it should be asserted and the caller fixed not to do so.

If it's OK to use both, then the logic should actually use both.

Am I missing something?

Perhaps you could show an example of where both are provided so that I it's easier to see why can't it be fixed in the caller?

I also realize that this part of PR perhaps doesn't really belong here. Should I split it off into a separate PR? I was just getting lost in the init logic for deepspeed and was trying to make it easier to follow.

Reply:
> Pre-this PR logic silently ignored model_init when model was provided.

No that is wrong, since at the beginning of training, `self.model_init` (if set) is called and replaces the model. So the previous logic is that the model was set, but overwritten at the beginning of training.

I agree that a user should not pass both, but suddenly raising an error for this is a breaking change, which we shouldn't do with a major release at least. That's why I suggested a warning, which tells the user something is probably wrong without breaking things.

I don't have an example in mind but since there is no hard error right now, there is probably code out there instantiating a `Trainer` with both a model and a model_init function, most likely for some hyper-parameter search since this is where `model_init` is the most useful.

Reply:
I totally support that - how do we implement a deprecation cycle in such a situation?

Reply:
We can raise a deprecation warning here instead of your error, and at the next major release switch to a hard error.

Reply:
Sounds good.

I'd be nice to have a doc somewhere at how to properly do deprecations in various situations.

Reply:
We simply use `warnings.warn(useful_message, FutureWarning)`.

Reply:
Yes, thank you - but it'd still be helpful to have a common format for the suggestion. I wrote:
```
                warnings.warn(
                    "`Trainer` requires either a `model` or `model_init` argument, but not both. "
                    "Ignoring `model_init`. This will become a fatal error in the next release",
                    FutureWarning,
                )
``` 
I'm not sure about the last sentence.